### PR TITLE
Don't run inspections on files outside a valid project

### PR DESCRIPTION
**Pull request by ortem**
Currently, our inspections run on any Rust files, including cases when Rust toolchain is not configured, or invalid or when a file does not belong to a Cargo project.

Now most of the inspections analyze only files that:
1. Have configured and valid toolchain
2. Belong to Cargo project and with workspace
3. Are included in the module tree

Special cases:
* Lightweight inspections (like syntax-only or naming lints) does not require (1)-(3)
* `RsDetachedFileInspection` does not require (3)

### CODE REVIEW

**Review by Undin — COMMENTED** [review]
I'm not sure that we want apply this change for all our inspections
It make sense when inspection interacts with project structure, stdlib, etc.
But there is a lot of very local inspections. For example, `RsWhileTrueLoopInspection` just checks `while (true)` constructions and suggest to replace it with `loop`

**Comment by ortem** [comment]
@Undin I've changed the implementation to be more flexible and generic, what do you think?

File: src/main/kotlin/org/rust/ide/inspections/RsDetachedFileInspection.kt
Comment (Kobzol) [inline src/main/kotlin/org/rust/ide/inspections/RsDetachedFileInspection.kt:66]: Is it needed to distinguish between project and workspace here, i.e. is there any inspection that needs that distinction?

Comment (ortem) [reply src/main/kotlin/org/rust/ide/inspections/RsDetachedFileInspection.kt:66]: That's a good question, I'm not sure about this separation too and just left the most general implementation for this review. Let's wait for a review from @vlad20012

Comment (Undin) [reply src/main/kotlin/org/rust/ide/inspections/RsDetachedFileInspection.kt:66]: I agree with @Kobzol and think that we should merge `Project` and `Workspace` because I hardly can come up with situation when these requirements may be different

**Comment by vlad20012** [comment]
Hmm, looks like requirements are not additive but inherent. I.e. a fale can't have a crate root without a workspace, can't have workspace without a cargo project, and can't have cargo project without a toolchain...

File: src/main/kotlin/org/rust/ide/inspections/RsDetachedFileInspection.kt
Comment (ortem) [reply src/main/kotlin/org/rust/ide/inspections/RsDetachedFileInspection.kt:66]: Thanks, I got rid of `Project` requirement

**Comment by ortem** [comment]
@vlad20012 @Undin  Do you think there should be just one requirement per inspection, but not a set of them?

**Comment by ortem** [comment]
@Undin Should I continue working on this PR and change something? Do you think it is useful or rather unnecessary?

**Comment by Undin** [comment]
@ortem I suppose this API is useful but looks like we don't need so complicated requirements. Maybe it's worth just using binary differentiation: inspections that require a fully configured project model (toolchain + workspace + crate root) or just syntax

**Comment by ortem** [comment]
@Undin I agree with you. Please review the new implementation based on the binary differentiation

File: src/main/kotlin/org/rust/ide/inspections/RsLocalInspectionTool.kt
Comment (vlad20012) [inline src/main/kotlin/org/rust/ide/inspections/RsLocalInspectionTool.kt:55]:
```suggestion
        return file.crate != null
            && file.project.toolchain?.looksLikeValidToolchain() == true
```

**Review by vlad20012 — APPROVED** [review]
bors r+

**Pull request merged**
